dart/optimizer/GradientDescentSolver.cpp

```diff
@@ -0,0 +1,482 @@
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <iostream>
+
+#include "dart/common/Console.h"
+#include "dart/optimizer/GradientDescentSolver.h"
+#include "dart/optimizer/Problem.h"
+
+namespace dart {
+namespace optimizer {
+
+//==============================================================================
+const std::string GradientDescentSolver::Type = "GradientDescentSolver";
+
+//==============================================================================
+GradientDescentSolver::UniqueProperties::UniqueProperties(
+    double _stepMultiplier,
+    size_t _maxAttempts,
+    size_t _perturbationStep,
+    double _maxPerturbationFactor,
+    double _maxRandomizationStep,
+    double _defaultConstraintWeight,
+    Eigen::VectorXd _eqConstraintWeights,
+    Eigen::VectorXd _ineqConstraintWeights)
+  : mStepSize(_stepMultiplier),
+    mMaxAttempts(_maxAttempts),
+    mPerturbationStep(_perturbationStep),
+    mMaxPerturbationFactor(_maxPerturbationFactor),
+    mMaxRandomizationStep(_maxRandomizationStep),
+    mDefaultConstraintWeight(_defaultConstraintWeight),
+    mEqConstraintWeights(_eqConstraintWeights),
+    mIneqConstraintWeights(_ineqConstraintWeights)
+{
+  // Do nothing
+}
+
+//==============================================================================
+GradientDescentSolver::Properties::Properties(
+    const Solver::Properties& _solverProperties,
+    const UniqueProperties& _descentProperties)
+  : Solver::Properties(_solverProperties),
+    UniqueProperties(_descentProperties)
+{
+  // Do nothing
+}
+
+//==============================================================================
+GradientDescentSolver::GradientDescentSolver(const Properties& _properties)
+  : Solver(_properties),
+    mGradientP(_properties),
+    mRD(),
+    mMT(mRD()),
+    mDistribution(0.0, std::nextafter(1.0, 2.0)) // This allows mDistrubtion to produce numbers in the range [0,1] inclusive
+{
+  // Do nothing
+}
+
+//==============================================================================
+GradientDescentSolver::GradientDescentSolver(std::shared_ptr<Problem> _problem)
+  : Solver(_problem),
+    mRD(),
+    mMT(mRD()),
+    mDistribution(0.0, std::nextafter(1.0, 2.0))
+{
+  // Do nothing
+}
+
+//==============================================================================
+GradientDescentSolver::~GradientDescentSolver()
+{
+  // Do nothing
+}
+
+//==============================================================================
+static double sign(double value)
+{
+  return value > 0? 1.0 : (value < 0? -1.0 : 0.0);
+}
```

Comment:
That implementation looks interesting. My only concern would be whether there's ever a case where returning an integer (instead of the original type) might be bad, but I can't think of a specific example.